### PR TITLE
contrib/neovide: new package (0.12.1)

### PR DESCRIPTION
**Pull request by subnut**
TODO
- [X] find source for gn with stable tarball SHA
- [ ] cross?
- [ ] hardening
- [ ] move to user/

### CODE REVIEW

File: contrib/neovide/template.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
+pkgname = "neovide"
+pkgver = "0.12.1"
+pkgrel = 0
+build_wrksrc = "neovide"
+build_style = "cargo"
+hostmakedepends = [
+    "cargo",
+    "ninja",
+    "libcxx-devel-static",
```
Comment (q66) [inline contrib/neovide/template.py:9]: ??

File: contrib/neovide/template.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+makedepends = [
+    "fontconfig-devel",
+    "freetype-devel",
+    "libexpat-devel",
+    "libjpeg-turbo-devel",
+    "libpng-devel",
+    "zlib-devel",
+]
+checkdepends = ["neovim"]
+pkgdesc = "No Nonsense Neovim Client in Rust"
```
Comment (q66) [inline contrib/neovide/template.py:21]: this violates pkgdesc conventions in at least 3 ways (and nothing about this seems no-nonsense)

File: contrib/neovide/template.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+maintainer = "Subhaditya Nath <sn03.general@gmail.com>"
+license = "MIT"
+url = "https://neovide.dev"
+_skia = (
```
Comment (q66) [inline contrib/neovide/template.py:25]: use individual variables

File: contrib/neovide/template.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+source = [
+    f"https://github.com/neovide/neovide/archive/refs/tags/{pkgver}.tar.gz>{pkgname}-{pkgver}.tar.gz",
+    f"https://gn.googlesource.com/gn/+archive/{_gn}.tar.gz>gn-{_gn[:8]}.tar.gz",
+    f"https://github.com/rust-skia/skia/archive/{_skia[0]}.tar.gz",
+    f"https://github.com/google/wuffs-mirror-release-c/archive/{_skia[1]}.tar.gz",
+    f"https://github.com/KhronosGroup/SPIRV-Cross/archive/{_skia[2]}.tar.gz",
+    f"https://github.com/GPUOpen-LibrariesAndSDKs/VulkanMemoryAllocator/archive/{_skia[3]}.tar.gz",
+]
+source_paths = [
+    "neovide",
```
Comment (q66) [inline contrib/neovide/template.py:41]: this can probably be . and then the build_wrksrc can be dropped

File: contrib/neovide/update.py [hunk 5]
```diff
@@ -0,0 +1,2 @@
+url = "https://github.com/neovide/neovide/tags"
```
Comment (q66) [inline contrib/neovide/update.py:1]: this should probably work automatically

File: contrib/neovide/template.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    from cbuild.util import cargo
+
+    with self.pushd(build_wrksrc):
+        self.cargo.vendor()
+        cargo.setup_vendor(self)
+
+
+def do_configure(self):
+    with self.pushd("../gn"):
+        self.do("python", "build/gen.py")
```
Comment (q66) [inline contrib/neovide/template.py:74]: use wrksrc

File: contrib/neovide/template.py (hunk 1 above)
Comment (nekopsykose) [reply contrib/neovide/template.py:9]: we could also make gn a separate template. they don't actually have breaking changes very often at all (i did that for a while for chromium and it built even with outdated stuff fine, up until rust was added), and all that would be needed is a source mirror (like github)

speaking of the gn.googlesource here is completely unusable because everything from googlesource is from gitiles and there are no stable tarballs (new sha every fetch)

File: contrib/neovide/template.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    "SKIA_GN_COMMAND": f"/builddir/{pkgname}-{pkgver}/gn/out/gn",
+    "SKIA_NINJA_COMMAND": "ninja",
+    "SKIA_USE_SYSTEM_LIBRARIES": "1",
+}
+
+
+def do_prepare(self):
+    from cbuild.util import cargo
+
+    with self.pushd(build_wrksrc):
```
Comment (q66) [inline contrib/neovide/template.py:67]: this shouldn't be needed with proper directory layout

File: contrib/neovide/template.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
+pkgname = "neovide"
+pkgver = "0.12.1"
+pkgrel = 0
+build_wrksrc = "neovide"
```
Comment (q66) [inline contrib/neovide/template.py:4]: this uses rustls so it should be in user/ repository and have restricted archs set

Comment (nekopsykose) [reply contrib/neovide/template.py:4]: rin builds on ppc64le/riscv64 since 0.17 however (this is 0.17.7). they merged the generic fallback

Comment (nekopsykose) [reply contrib/neovide/template.py:4]: (i.e. this is not restricted because of rustls at all. maybe other reasons..)

File: contrib/neovide/template.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+def pre_build(self):
+    with self.pushd("../gn"):
+        self.do("ninja", "-C", "out")
+
+
+def do_install(self):
+    self.install_license("LICENSE")
+    self.install_bin("target/*-chimera-linux-*/release/neovide", glob=True)
+    self.install_file("assets/neovide.desktop", "usr/share/applications")
+    for f in self.find("assets", "neovide-*.png"):
```
Comment (nekopsykose) [inline contrib/neovide/template.py:86]: there is an svg too

File: contrib/neovide/template.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+
+def pre_build(self):
+    with self.pushd("../gn"):
+        self.do("ninja", "-C", "out")
+
+
+def do_install(self):
+    self.install_license("LICENSE")
+    self.install_bin("target/*-chimera-linux-*/release/neovide", glob=True)
```
Comment (nekopsykose) [inline contrib/neovide/template.py:84]: self.profile().triplet instead of glob

File: contrib/neovide/template.py (hunk 1 above)
Comment (subnut) [reply contrib/neovide/template.py:9]: Ah that's why the CI failed with mismatching SHA.

Didn't know about that.

File: contrib/neovide/update.py [hunk 11]
```diff
@@ -0,0 +1,2 @@
...
+pattern = r'/archive/refs/tags/([\d.]+)(?=\.tar\.gz")'
```
Comment (nekopsykose) [inline contrib/neovide/update.py:2]:
```suggestion
pattern = r'/archive/refs/tags/([\d.]+).tar.gz'
```

you don't need fancy captures

File: contrib/neovide/template.py (hunk 8 above)
Comment (q66) [reply contrib/neovide/template.py:4]: yeah but it's still statically linked tls so i don't want it in primary repos regardless

Comment (q66) [reply contrib/neovide/template.py:4]: archs should still be restricted because skia does not support big endian

File: contrib/neovide/template.py (hunk 2 above)
Comment (subnut) [reply contrib/neovide/template.py:21]: oh wait this was written by past me

will fix

---

**TL;DR** yesterday i thought of writing a template for this package, and today I discovered that a template already exists! So I got to work trying to fix it's build without even looking at the rest of the stuff.

That's why I was initially building 0.11.2 instead of the latest 0.12.1 (if you saw that on IRC)

File: contrib/neovide/update.py (hunk 5 above)
Comment (subnut) [reply contrib/neovide/update.py:1]: nah the rest of the URLs clobber it for some reason.

will try again tho.

File: contrib/neovide/template.py (hunk 8 above)
Comment (nekopsykose) [reply contrib/neovide/template.py:4]: yeah, written as `if self.profile().endian == "big": broken=`

File: contrib/neovide/template.py (hunk 9 above)
Comment (subnut) [reply contrib/neovide/template.py:86]: where should that go?

File: contrib/neovide/template.py (hunk 8 above)
Comment (q66) [reply contrib/neovide/template.py:4]: no, written as archs= because skia quite literally has arch-specific code in it

File: contrib/neovide/template.py (hunk 9 above)
Comment (subnut) [reply contrib/neovide/template.py:86]: oh icons/hicolor/scalable/apps/ okay

File: contrib/neovide/update.py (hunk 11 above)
Comment (subnut) [reply contrib/neovide/update.py:2]: sorry... idk anything about the python-specific regex stuff  
(skill issue basically)

will fix.

File: contrib/neovide/patches/fix-skia-bindings.patch [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+-                "obj/third_party/icu/icu.ninja".into(),
+                 "obj/modules/skunicode/skunicode.ninja".into(),
+             ]);
++            if !env::use_system_libraries() {
++                files.push("obj/third_party/icu/icu.ninja".into());
++            }
+         }
+         if features.svg {
+             files.push("obj/modules/svg/svg.ninja".into());
+--- a/neovide/vendor/skia-bindings/.cargo-checksum.json
```
Comment (q66) [inline contrib/neovide/patches/fix-skia-bindings.patch:23]: unneeded/unwanted, use cargo.clear_vendor_checksums

Comment (q66) [reply contrib/neovide/patches/fix-skia-bindings.patch:23]: also i'm about 95% sure you'll need to further patch skia (see ppc-musttail.patch in firefox/thunderbird, similar in libreoffice and chromium)

File: contrib/neovide/update.py (hunk 5 above)
Comment (subnut) [reply contrib/neovide/update.py:1]: Nah it doesn't seem to be working

without `update.py` -
```
$ cbuild update-check contrib/neovide verbose
Checking for updates: neovide=0.12.1
Adding 'https://neovide.dev' for version check...
Adding 'https://github.com/neovide/neovide/archive/refs/tags/' for version check...
Adding 'https://gn.googlesource.com/gn/+archive/' for version check...
Adding 'https://github.com/rust-skia/skia/archive/' for version check...
Adding 'https://github.com/google/wuffs-mirror-release-c/archive/' for version check...
Adding 'https://github.com/KhronosGroup/SPIRV-Cross/archive/' for version check...
Adding 'https://github.com/GPUOpen-LibrariesAndSDKs/VulkanMemoryAllocator/archive/' for version check...
Fetching 'https://neovide.dev' for version checks...
Fetching 'https://github.com/neovide/neovide/tags' for version checks...
Fetching 'https://gn.googlesource.com/gn/+archive/' for version checks...
Fetching 'https://github.com/rust-skia/skia/tags' for version checks...
Fetching 'https://github.com/google/wuffs-mirror-release-c/tags' for version checks...
Fetching 'https://github.com/KhronosGroup/SPIRV-Cross/tags' for version checks...
Fetching 'https://github.com/GPUOpen-LibrariesAndSDKs/VulkanMemoryAllocator/tags' for version checks...
Checking found version: 0.3.0
Checking found version: 0.3.1
Checking found version: 0.3.2
Checking found version: 0.3.3
Checking found version: 0.10.0
Checking found version: 0.10.1
Checking found version: 0.10.2
Checking found version: 0.10.3
Checking found version: 0.10.4
Checking found version: 0.11.0
Checking found version: 0.11.1
Checking found version: 0.11.2
Checking found version: 0.12.0
Checking found version: 0.12.1
Checking found version: 1.0.0
Checking found version: 1.0.1
Checking found version: 2.0.0
Checking found version: 2.1.0
Checking found version: 2.2.0
Checking found version: 2.3.0
Checking found version: 3.0.0
Checking found version: 3.0.1
contrib/neovide: 0.12.1 -> 1.0.0
contrib/neovide: 0.12.1 -> 1.0.1
contrib/neovide: 0.12.1 -> 2.0.0
contrib/neovide: 0.12.1 -> 2.1.0
contrib/neovide: 0.12.1 -> 2.2.0
contrib/neovide: 0.12.1 -> 2.3.0
contrib/neovide: 0.12.1 -> 3.0.0
contrib/neovide: 0.12.1 -> 3.0.1
```

File: contrib/neovide/template.py (hunk 2 above)
Comment (subnut) [reply contrib/neovide/template.py:21]: "Graphical client for neovim" sounds ok?

File: contrib/neovide/template.py (hunk 8 above)
Comment (subnut) [reply contrib/neovide/template.py:4]: @q66 what should be the value of `archs =` ?

(I am a complete noob when it comes to architectures)

File: contrib/neovide/update.py (hunk 11 above)
Comment (subnut) [reply contrib/neovide/update.py:2]: wait shouldn't it be `\.tar\.gz` ?

Comment (nekopsykose) [reply contrib/neovide/update.py:2]: there's no realistic case where the . matters for being dot-only or any

**Comment by subnut** [comment]
Some changes made, some yet to be made.

**Comment by subnut** [comment]
> yeah but it's still statically linked tls so i don't want it in primary repos regardless

@q66 if i replaced rustls with tls-native (i.e. openssl), would this package be accepted into `contrib` ?

File: contrib/neovide/template.py (hunk 1 above)
Comment (subnut) [reply contrib/neovide/template.py:9]: switched to koji.fedoraproject.org for gn

File: contrib/neovide/patches/fix-skia-bindings.patch (hunk 12 above)
Comment (subnut) [reply contrib/neovide/patches/fix-skia-bindings.patch]: added ppc-musttail.patch from firefox
